### ISSUE TITLE
Default initialization does not match default nonlinearity

### ISSUE THREAD

**Issue by nlessmann** [Lasagne/Lasagne#658]
For example, the Conv2DLayer has the following default parameters:

    W=init.GlorotUniform(),
    nonlinearity=nonlinearities.rectify, ...

But the default gain for the Glorot initialization is 1.0, which is optimal for sigmoids and not ReLUs. Shouldn't the layer by default be initialized with Glorot initialization with a gain of sqrt(2) if the default nonlinearities are ReLUs?

**Comment by benanne** [comment]
That's a good point. However, people are much more likely to change the nonlinearity of the layer than the initialization strategy, so I'm not sure if that would be a good idea. The nice thing about keeping the default gain at `1.0` is that it actually works fine for almost all possible nonlinearities, unless you go beyond 20-30 layers. It takes a while for a product of missing gain factors `sqrt(2)` to accumulate into a problematic scaling constant that spans several order of magnitude (`sqrt(2)**20` is only about `10**3`).

In principle you are right though, they are not matched. Unfortunately changing defaults like these is rather difficult, as it is likely to break a lot of people's old code. It would take a very long deprecation period and I'm not sure it's worth it. If we do ever decide to change some of these, I would also advocate for replacing `GlorotUniform` with `Orthogonal`.

Sidenote: I don't think a gain of `1.0` is optimal for the sigmoid nonlinearity. In fact, for any sigmoidal function (including tanh), the optimal gain actually depends on the distribution you sample from. Yann LeCun and colleagues used a "scaled tanh" that incorporated scaling factors to bring the optimal gain to 1.0 in some of their early work, iirc. If you're interested in this topic, there was an interesting discussion on Google+ a while back that a bunch of researchers (including myself) participated in:
first part: https://plus.google.com/+EricBattenberg/posts/f3tPKjo7LFa
second part: https://plus.google.com/+SoumithChintala/posts/RZfdrRQWL6u

**Comment by f0k** [comment]
I agree with Sander, for backwards compatibility we cannot easily change the defaults. If you train very deep networks and need to be careful about the scale, note that data-dependent initialization is not very difficult to implement: http://arxiv.org/abs/1511.06422
This may even allow you to limit batch normalization to correcting the mean only: http://arxiv.org/abs/1602.07868

**Comment by felixlaumon** [comment]
I always find the defaults strange, and I actually changed the defaults in my local clone of Lasagne. Maybe there should be some comments about the need to change the init when changing the nonlinearity.

**Comment by f0k** [comment]
> I always find the defaults strange, and I actually changed the defaults in my local clone of Lasagne.

We could think about adding some way to change the defaults from a script, as we discussed for the nonlinearities (#138). On the other hand, this will probably not be a lot more compact than just setting everything explicitly, such as:
```python
kws = {'nonlinearity': lasagne.nonlinearities.LeakyRectify(.1),
       'init': lasagne.init.Orthogonal(gain=np.sqrt(2/(1+.1**2)))}
layer = InputLayer(...)
layer = Conv2DLayer(..., **kws)
layer = ...
layer = DenseLayer(..., **kws)
```

> Maybe there should be some comments about the need to change the init when changing the nonlinearity.

Where would you look for this? Adding this to the docstring for every `W` param may be a bit overkill (would you have seen it there?). Maybe the best way would be continuing the Lasagne tutorial, to explain what needs to be taken care of when going very deep (among other things that need further explanation: binary classification, regression, multi-input networks, multi-output networks).

**Comment by felixlaumon** [comment]
As crazy as it sounds, have you guys considered having fewer defaults and requiring the user the define every important parameters?

> Where would you look for this? Adding this to the docstring for every W param may be a bit overkill (would you have seen it there?)

On second thought, this is a very good point. The comments should be concise and not become a textbook about neural network.

**Comment by f0k** [comment]
> As crazy as it sounds, have you guys considered having fewer defaults and requiring the user the define every important parameters?

Yes, please check the long discussion in #138. There have been proponents of "explicit is better than implicit" and proponents of making it easier / avoiding pitfalls for novices. In #180 we could have changed the default to use `gain='relu'`, but either nobody thought of it or we didn't want to break backwards compatibility.

Note that you are free to define every important parameter yourself even if there are defaults. That's partly why the "explicit is better than implicit" fraction has given in -- you can still make your code explicit, we're just not forcing it on everybody. That sounded convincing enough back then.
